File: services/rag/src/embeddings/openai.ts
import type { EmbeddingClient, EmbeddingConfig } from "./types.js";

export class OpenAIEmbedding implements EmbeddingClient {
  private readonly apiKey: string;
  readonly model: string;
  readonly dimensions: number;
  private readonly baseUrl = "https://api.openai.com/v1";

  constructor(config: EmbeddingConfig) {
    this.apiKey = config.apiKey;
    this.model = config.model;
    this.dimensions = config.dimensions ?? 1536;
  }

  async embed(texts: string[]): Promise<number[][]> {
    const response = await fetch(`${this.baseUrl}/embeddings`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${this.apiKey}`,
      },
      body: JSON.stringify({
        model: this.model,
        input: texts,
        dimensions: this.dimensions,
      }),
    });

    if (!response.ok) {
      const error = await response.text();
      throw new Error(`OpenAI embedding failed: ${error}`);
    }

    const data = (await response.json()) as {
      data: Array<{ index: number; embedding: number[] }>;
    };
    const sorted = [...data.data].sort((a, b) => a.index - b.index);
    return sorted.map((item) => item.embedding);
  }

  async embedSingle(text: string): Promise<number[]> {
    const [embedding] = await this.embed([text]);
    return embedding;
  }
}

File: services/eval/src/rag_eval/embeddings/openai.py
from typing import Self

import httpx

from rag_eval.http_retry import post_with_retry

DEFAULT_DIMENSIONS = 1536


class OpenAIQueryEmbedding:
    def __init__(
        self,
        api_key: str,
        model: str,
        dimensions: int | None = None,
        base_url: str = "https://api.openai.com/v1",
        client: httpx.Client | None = None,
    ) -> None:
        self._api_key = api_key
        self.model = model
        self.dimensions: int | None = dimensions if dimensions is not None else DEFAULT_DIMENSIONS
        self._base_url = base_url
        self._client = client or httpx.Client(timeout=30.0)

    def embed_query(self, text: str) -> list[float]:
        data = post_with_retry(
            self._client,
            f"{self._base_url}/embeddings",
            service="OpenAI embedding",
            json={"model": self.model, "input": [text], "dimensions": self.dimensions},
            headers={"Authorization": f"Bearer {self._api_key}"},
        )
        return data["data"][0]["embedding"]

    def close(self) -> None:
        self._client.close()

    def __enter__(self) -> Self:
        return self

    def __exit__(self, *exc_info: object) -> None:
        self.close()
